### PR TITLE
implement `aes_encrypt/aes_decrpyt/random_bytes`

### PR DESCRIPTION
**Pull request by iosmanthus**
- add `aes_encrypt/aes_decrpyt` via `OpenSSL`
`ecb/cbc/cfb1/cfb8/cfb128/ofb` mode
- add `random_bytes`

Signed-off-by: iosmanthus <myosmanthustree@gmail.com>

## What have you changed? (mandatory)

1. Use OpenSSL just like: https://github.com/tikv/tikv/pull/3886, add `aes_encrypt/aes_decrpyt`, the block encryption mode can be specified by an [ `Nid` ](https://docs.rs/openssl/0.10.20/openssl/nid/struct.Nid.html), now it's fully MySQL compatible supporting `ecb/cbc/cfb1/cfb8/cfb128/ofb`
2. Add `random_byte` which can be used to generate `IV`/`key` for `aes_encrypt/aes_decrpyt`

## What are the type of changes? (mandatory)
- New features (a change which adds functionality)

## How has this PR been tested? (mandatory)
`cargo test builtin_encryption`

### CODE REVIEW

File: src/coprocessor/dag/expr/builtin_encryption.rs [hunk 1]
```diff
@@ -21,6 +25,62 @@ const SHA256: i64 = 256;
...
+
+/// Generate a sequence of random bytes in length of `len`
+/// TODO: upgrade crate `rand` to 0.6+
+/// with `rand` 0.6+, the code can be written:
+/// ```compile_fail
+/// Standard.sample_iter(&mut thread_rng()).take(len).collect()
+/// ```
+fn random_bytes(len: usize) -> Vec<u8> {
+    let mut rng = rand::thread_rng();
+    (0..len).map(|_| rng.gen()).collect()
```
Comment (kennytm) [inline src/coprocessor/dag/expr/builtin_encryption.rs]: With all versions of `rand` from 0.3 to 0.6 you could simply write

```suggestion
    let mut bytes = vec![0u8; len];
    rng.fill_bytes(&mut bytes);
    bytes
```

This is much faster even considering the zero-initialization. With rand 0.6.5:

```
test bench_fill_bytes_long        ... bench:       1,068 ns/iter (+/- 103)
test bench_fill_bytes_short       ... bench:          45 ns/iter (+/- 6)
test bench_loop_and_collect_long  ... bench:       7,407 ns/iter (+/- 1,157)
test bench_loop_and_collect_short ... bench:          85 ns/iter (+/- 10)
```

With rand 0.3.14:

```
test bench_fill_bytes_long        ... bench:       3,834 ns/iter (+/- 208)
test bench_fill_bytes_short       ... bench:          59 ns/iter (+/- 2)
test bench_loop_and_collect_long  ... bench:      17,345 ns/iter (+/- 1,747)
test bench_loop_and_collect_short ... bench:         164 ns/iter (+/- 7)
```

<details><summary>Benchmark</summary>

```rust
#![feature(test)]

extern crate test;
extern crate rand;

use test::{Bencher, black_box};
use rand::{Rng, RngCore, thread_rng};

#[bench]
fn bench_loop_and_collect_long(b: &mut Bencher) {
	let mut rng = thread_rng();
	let len = 2048;
	b.iter(|| {
		black_box((0..black_box(len)).map(|_| rng.gen()).collect::<Vec<u8>>());
	});
}

#[bench]
fn bench_fill_bytes_long(b: &mut Bencher) {
        let mut rng = thread_rng();
        let len = 2048;
        b.iter(|| {
		let mut bytes = vec![0u8; black_box(len)];
		rng.fill_bytes(&mut bytes);
		black_box(bytes);
        });
}

#[bench]
fn bench_loop_and_collect_short(b: &mut Bencher) {
        let mut rng = thread_rng();
        let len = 16;
        b.iter(|| {
                black_box((0..black_box(len)).map(|_| rng.gen()).collect::<Vec<u8>>());
        });
}

#[bench]
fn bench_fill_bytes_short(b: &mut Bencher) {
        let mut rng = thread_rng();
        let len = 16;
        b.iter(|| {
                let mut bytes = vec![0u8; black_box(len)];
                rng.fill_bytes(&mut bytes);
                black_box(bytes);
        });
}
```

</details>

**Comment by siddontang** [comment]
oh, @breeswish why https://github.com/tikv/tikv/pull/3886 has not been reviewed for so long time :sob:

**Comment by iosmanthus** [comment]
It seems some block cipher modes are not supported in `CI` which lead to an errror while try to create a cipher  from `mode`.However, I passed all the unit test in `coprocessor::dag::expr::builtin_encryption::tests` with `OpenSSL (version 1.1.1)`.

**Comment by iosmanthus** [comment]
> It seems some block cipher modes are not supported in `CI` which lead to an error while try to create a cipher from `mode`. However, I passed all the unit tests in `coprocessor::dag::expr::builtin_encryption::tests` with `OpenSSL (version 1.1.1)` locally.

I am totally wrong.. 😢. If `OpenSSL` is `1.0.2 or below` it need to be explicitly initialized...

File: src/coprocessor/dag/expr/builtin_encryption.rs [hunk 2]
```diff
@@ -21,6 +25,59 @@ const SHA256: i64 = 256;
...
+    653i32, // aes-128-cfb8
+    654i32, // aes-192-cfb8
+    655i32, // aes-256-cfb8
+};
+
+/// Generate a sequence of random bytes in length of `len`
+fn random_bytes(len: usize) -> Vec<u8> {
+    let mut rng = rand::thread_rng();
+    let mut bytes = vec![0; len];
+    rng.fill_bytes(&mut bytes);
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_encryption.rs]: could we use `rand::thread_rng().fill_bytes(&mut bytes)` directly?

File: src/coprocessor/dag/expr/builtin_encryption.rs [hunk 3]
```diff
@@ -164,14 +221,86 @@ impl ScalarFunc {
...
+
+    pub fn aes_encrypt<'a, 'b: 'a>(
+        &'b self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let plaintext = try_opt!(self.children[0].eval_string(ctx, row));
+        let key = try_opt!(self.children[1].eval_string(ctx, row));
+        let iv = try_opt!(self.children[2].eval_string(ctx, row));
+        let mode = try_opt!(self.children[3].eval_int(ctx, row)) as i32;
```
Comment (AndreMouche) [inline src/coprocessor/dag/expr/builtin_encryption.rs]: e, it seems different from tidb's implementation @lonng could you help to review it?

Comment (AndreMouche) [reply src/coprocessor/dag/expr/builtin_encryption.rs]: @zz-jason Could you help to review it?

**Comment by iosmanthus** [comment]
The version of `OpenSSL` in CI is 1.0.2 or below, it needs to be explicitly initialized... I just initialize it in unit tests but not globally... Maybe we need a PR that imports `OpenSSL` and initializes it globally.

File: Cargo.toml [hunk 4]
```diff
@@ -122,6 +122,8 @@ vlog = "0.1.4"
 twoway = "0.2.0"
 ordered-float = "1.0"
 nom = "5.0.0-beta1"
+openssl = { version = "0.10.23", features = ["vendored"] }
```
Comment (breeswish) [inline Cargo.toml]: why are we using replace, instead of directly specify libre-ssl as the dependency?

**Comment by BusyJay** [comment]
This is not going to work as expected until pingcap/grpc-rs#326 is merged.

**Comment by overvenus** [comment]
PTAL @hicqu , should we close this PR?

**Comment by overvenus** [comment]
PTAL @iosmanthus , should we close this PR?

**Comment by iosmanthus** [comment]
> PTAL @iosmanthus , should we close this PR?

@overvenus Jay has already merged https://github.com/pingcap/grpc-rs/pull/326 , I will handle the conflicts!

File: src/coprocessor/dag/expr/builtin_encryption.rs [hunk 5]
```diff
@@ -21,6 +25,59 @@ const SHA256: i64 = 256;
 const SHA384: i64 = 384;
 const SHA512: i64 = 512;
 
+/// Numerical identifiers for an OpenSSL object supported in TiKV.
+const NID: phf::Set<i32> = phf_set! {
```
Comment (breeswish) [inline src/coprocessor/dag/expr/builtin_encryption.rs]: I'm curious why do we need a phf here? Is it possible to only use openssl?

**Comment by lonng** [comment]
Ping @iosmanthus

**Comment by BusyJay** [comment]
Now that the ssl library is vendored, I think you can continue to work on this.

**Comment by iosmanthus** [comment]
> Now that the ssl library is vendored, I think you can continue to work on this.

OK! I will try to migrate them to RPN framework too.

**Pull request closed without merging**